share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManagerTest.java

```diff
@@ -0,0 +1,209 @@
+    private static final SharePartitionKey KEY2 = SharePartitionKey.getInstance("gs2", Uuid.randomUuid(), 0);
+    private static final SharePartitionKey KEY3 = SharePartitionKey.getInstance("gs1", Uuid.randomUuid(), 1);
+    private static final SharePartitionKey KEY4 = SharePartitionKey.getInstance("gs1", Uuid.randomUuid(), 7);
+
+    @BeforeEach
+    public void setUp() {
+        manager = new ShareCoordinatorOffsetsManager(new SnapshotRegistry(new LogContext()));
+    }
+
+    @Test
+    public void testUpdateStateAddsToInternalState() {
+        manager.updateState(KEY1, 0L);
+        assertEquals(Optional.empty(), manager.lastRedundantOffset());
+
+        manager.updateState(KEY1, 10L);
+        assertEquals(Optional.of(10L), manager.lastRedundantOffset()); // [0-9] offsets are redundant.
+
+        manager.updateState(KEY2, 15L);
+        assertEquals(Optional.empty(), manager.lastRedundantOffset());  // No update to last redundant after adding 15L so, empty.
+
+        assertEquals(10L, manager.curState().get(KEY1));
+        assertEquals(15L, manager.curState().get(KEY2));
+    }
+
+    private static class ShareOffsetTestHolder {
+        static class TestTuple {
+            final SharePartitionKey key;
+            final long offset;
+            final Optional<Long> expectedOffset;
+
+            private TestTuple(SharePartitionKey key, long offset, Optional<Long> expectedOffset) {
+                this.key = key;
+                this.offset = offset;
+                this.expectedOffset = expectedOffset;
+            }
+
+            static TestTuple instance(SharePartitionKey key, long offset, Optional<Long> expectedOffset) {
+                return new TestTuple(key, offset, expectedOffset);
+            }
+        }
+
+        private final String testName;
+        private final List<TestTuple> tuples;
+        private final boolean shouldRun;
+
+        ShareOffsetTestHolder(String testName, List<TestTuple> tuples) {
+            this(testName, tuples, true);
+        }
+
+        ShareOffsetTestHolder(String testName, List<TestTuple> tuples, boolean shouldRun) {
+            this.testName = testName;
+            this.tuples = tuples;
+            this.shouldRun = shouldRun;
+        }
+    }
+
+    static Stream<ShareOffsetTestHolder> generateNoRedundantStateCases() {
+        return Stream.of(
+            new ShareOffsetTestHolder(
+                "no redundant state single key",
+                List.of(
+                    ShareOffsetTestHolder.TestTuple.instance(KEY1, 10L, Optional.empty())
+                )
+            ),
+
+            new ShareOffsetTestHolder(
+                "no redundant state multiple keys",
+                List.of(
+                    ShareOffsetTestHolder.TestTuple.instance(KEY1, 10L, Optional.empty()),
+                    ShareOffsetTestHolder.TestTuple.instance(KEY4, 11L, Optional.empty()),
+                    ShareOffsetTestHolder.TestTuple.instance(KEY2, 13L, Optional.empty())
```

Comment:
Small efficiency gain here as the offsets will be applied in increasing order (partition records auto inc offset) to the `updateState` method. If the case is such that 10L is the smallest one - it means there are no offsets smaller than that present in the topic partition, hence returning 10L will be of no consequence and save an extra `deleteRecords` call.

In the algorithm we have chosen to get at least 2 offsets for any key before exposing the redundant offset. 

Also since we are exposing the offset only once and then setting the boolean flag, the 2nd and 3rd line should be `empty`.